### ISSUE TITLE
Block response size check is not correct

### ISSUE THREAD

**Issue by nazar-pc** [paritytech/polkadot-sdk#1232]
I was re-reviewing my PR https://github.com/paritytech/substrate/pull/12146 and noticed that the size calculation (both before and after) is likely incorrect and can cause unexplainable errors.

Here estimation of total size accounts for some fields of `block_data`, but not others:
https://github.com/paritytech/polkadot-sdk/blob/0400ed90757d125f7769f49044632dac6b654cc4/substrate/client/network/sync/src/block_request_handler.rs#L419-L430

I'm not aware of what are the historical reasons for this, but if my understanding is correct, response will always be strictly bigger than `total_size` estimated. The message, annoyingly, is in protobuf format, so suspect there is no `encoded_size` method there, but I think it should push an entry into `blocks`, check encoded size and in case size is exceeded pop the value before breaking the loop.

**Comment by altonen** [comment]
Requests/responses are encoded using unsigned-varint which just prepends a length (at most 9 bytes) to the message. This message length tag is not counted in so the block response has all of the 16 MB to use.

It looks like right now the length is not checked for outbound messages which could cause "unexplainable errors" but I don't think exposing the details of the encoding should be exposed to the block request protocol and instead it should try to create the response within the limits that were provided when the protocol was initialized.

**Comment by nazar-pc** [comment]
If I understand correctly, we're encoding `BlockData`, while only checking the size of `body` and `indexed_body` fields, ignoring all others. This is where I think the problem is, every field and length for each vector should be accounted for.

**Comment by altonen** [comment]
Fair enough. I think all the information is available to `BlockRequestHandler` though so I don't think there is a need to expose any details about the implementation of the request-response protocol when crafting the response and the handler should just calculate the response size correctly based on the information it has.
